ee/server/service/devices.go

```diff
@@ -166,3 +172,73 @@ func (svc *Service) GetFleetDesktopSummary(ctx context.Context) (fleet.DesktopSu
 
 	return sum, nil
 }
+
+func (svc *Service) TriggerLinuxKeyEscrow(ctx context.Context, host *fleet.Host) error {
+	level.Debug(svc.logger).Log("msg", "trigger linux key escrow", "host_id", host.ID)
+
+	if !slices.Contains(LUKSSupportedPlatforms, host.Platform) {
+		return &fleet.BadRequestError{Message: "Host platform does not support key escrow"}
+	}
+
+	ac, err := svc.ds.AppConfig(ctx)
+	if err != nil {
+		return err
+	}
+	if !ac.MDM.EnabledAndConfigured {
+		return fleet.ErrMDMNotConfigured
+	}
+
+	// Is AppConfig.MDM.EnableDiskEncryption the disk encryption setting for “No team”, while
+	// TeamMDM.EnableDiskEncryption is for all other teams?
+	// confirm No team taemId
+	if host.TeamID == nil {
+		// No team
+		if !ac.MDM.EnableDiskEncryption.Value {
+			return &fleet.BadRequestError{Message: "Disk encryption is not enabled for this host"}
+		}
+	} else {
+		// Team
+		tc, err := svc.ds.TeamMDMConfig(ctx, *host.TeamID)
+		if err != nil {
+			return err
+		}
+		if !tc.EnableDiskEncryption {
+			return &fleet.BadRequestError{Message: "Disk encryption is not enabled for this host's team"}
+		}
+	}
+
+	_, err = svc.ds.IsHostConnectedToFleetMDM(ctx, host)
+	if err != nil {
+		return ctxerr.Wrap(ctx, err, "Host is not connected to Fleet MDM")
+	}
+
+	if !*host.DiskEncryptionEnabled {
+		return &fleet.BadRequestError{Message: "Host's disk is not encrypted. Please enable disk encryption for this host."}
+	}
+
+	if fleet.CompareVersions(*host.OrbitVersion, minOrbitLUKSVersion) == -1 {
+		return &fleet.BadRequestError{Message: "Host's Orbit version does not support this feature. Please upgrade Orbit to the latest version."}
+	}
+
+	// TODO (confirm details): put all of these errors into the
+	// `host_disk_encryption_keys.client_error` column
+	// Use ds.SetOrUpdateHostDiskEncryptionKey ?
+
+	key, err := svc.ds.GetHostDiskEncryptionKey(ctx, host.ID)
```

Comment:
Do we _want_ to bail when a key is set (which would disallow re-escrowing a key if the user has rotated their LUKS key)? Given that the old key will stay put until the new key is re-escrowed, it seems safe to let users request re-escrow when there's already a key in the DB.

Reply:
We do want to bail. Resolving this.